### PR TITLE
Implement SimplifyReturnExpressionToConstantTypeRector

### LINKED ISSUE #8262: SimplifyUselessVariableRector missing opportunity
# Bug Report

I expected SimplifyUselessVariableRector  to cleanup this code

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | e.g. v0.11.5 (invoke `vendor/bin/rector --version`)             |

## Minimal PHP Code Causing Issue

https://getrector.com/demo/b2fd9fbe-1209-4cbb-a317-fdd49ad2f2b1

## Expected Behaviour

```diff
<?php

function apcInstalled(): bool
{
    $apcInstalled = false;
-   $apcInstalled = false;
    if (rand(0,1)) {
        return true;
    }

-    return $apcInstalled;
+    return false;
}
```

### PR DESCRIPTION
**Pull request by staabm**
implemented the idea described in https://github.com/rectorphp/rector/issues/8262#issuecomment-1761351264

closes https://github.com/rectorphp/rector/issues/8262

before this PR we did not touch these returns:
https://getrector.com/demo/7ae0a195-0f84-4b56-9395-0d5214133819

### CODE REVIEW

File: rules-tests/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector/Fixture/multiple_modified_returns.php.inc [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+-----
+<?php
+
+namespace Rector\Tests\CodeQuality\Rector\Return_\SimplifyReturnExpressionToConstantTypeRector\Fixture;
+
+class MultipleModifiedReturns
+{
+    public function run()
+    {
+        $return = false;
```
Comment (staabm) [inline rules-tests/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector/Fixture/multiple_modified_returns.php.inc:35]: I intentionally don't clean up stray variables.. I think those will be cleaned up by a code-style tool

**Comment by staabm** [comment]
I would love a first round of feedback, whether the idea itself is acceptable

File: rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+ * @see \Rector\Tests\CodeQuality\Rector\Return_\SimplifyReturnExpressionToConstantTypeRector\SimplifyReturnExpressionToConstantTypeRectorTest
+ */
+final class SimplifyReturnExpressionToConstantTypeRector extends AbstractRector
+{
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('"something()" will be renamed to "somethingElse()"', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeClass
```
Comment (staabm) [inline rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php:25]: todo

**Comment by samsonasik** [comment]
I am against moving contains of variable  as may cause invalid overridden data internally, per discussed on https://github.com/rectorphp/rector/issues/8262#issuecomment-1761355374

that  can be something like ref https://3v4l.org/OVDTZ

that will make we "re-create" a variant of "MoveVariableDeclarationNearReferenceRector" which removed as buggy while already cover many things :)

- https://github.com/rectorphp/rector-src/pull/587

**Comment by staabm** [comment]
> that can be something like ref [3v4l.org/OVDTZ](https://3v4l.org/OVDTZ)

thanks for these tests - these were added and fixed easily.

> that will make we "re-create" a variant of "MoveVariableDeclarationNearReferenceRector" which removed as buggy while already cover many things :)

I don't think so. the rector-rule here is way simpler then "MoveVariableDeclarationNearReferenceRector" and does not at all move statements/variables arround.

**Comment by staabm** [comment]
@TomasVotruba whats your thoughts on this one?

**Comment by TomasVotruba** [comment]
I'm briefly looking into the rule and not sure what exactly it does. Got any PR on open-source project where you run such rule? :) That would be easy for me to understand. 

Replacing variables with expression can lead to side effect for various reason, so I'd avoid that by default in any rule.

As for the enum part, that makes sense :+1: Maybe it could be a single rule for enum support.

**Comment by staabm** [comment]
see the rule applied on one of our projects: https://github.com/complex-gmbh/kunzmann/pull/4943/commits/31297a04040b1ccd6bed88ae030535129aa358bd

File: rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkip($node) || $node->expr === null) {
+            return null;
+        }
+
+        $nativeType = $this->nodeTypeResolver->getNativeType($node->expr);
+        if ($nativeType instanceof ConstantScalarType) {
+            $constantValue = $nativeType->getValue();
```
Comment (samsonasik) [inline rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php]: is this work for `++$variable` or `extract()` before return?

I stil think this rule will be buggy, as variable can be modified or replaced magically on the fly before returning.

Comment (staabm) [reply rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php]: No. Its just the common path. 

I determine whether the type of a return expression is constant, if so replace the return expression with the constant (only true/false/null/enum-case)

Comment (samsonasik) [reply rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php]: that need to be added to test fixture imo, then, probably better to be handled via `StmtsAwareInterface` node.

The code like this should be skipped:

1. increment/decrement/postcrement/precrement

```php
$var = 1;

if (rand(0, 1)) {
   return true;
}

++$var;

return $var;
```

2. replaced by `extract()` magically

```php
$var = false;

if (rand(0, 1)) {
   return true;
}

extract($data);

return $var;
```

Comment (staabm) [reply rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php]: Let me add this tests, thanks

Comment (staabm) [reply rules/CodeQuality/Rector/Return_/SimplifyReturnExpressionToConstantTypeRector.php]: added. thx for the cases.

btw, I learned that incrementing `false` stays `false`:
https://3v4l.org/h2NC5

**Comment by samsonasik** [comment]
Could you run :

```
bin/rector
```

and verify if there is regression with this new rule?

**Comment by staabm** [comment]
@samsonasik excellent suggestion.

running the new rule on the rector-src codebase actually showed a few places which unnecessarily use a variable, even though they return always a constant value

- https://github.com/rectorphp/rector-src/pull/5167/commits/ec64beca9f2c8848d15f9ad1835d7702390c13ce#diff-2e4e369fc620d62a8e8cf0d62fdb790bed20fc22ac23b2ee4e6a5eb56d98cbd0
- https://github.com/rectorphp/rector-src/pull/5167/commits/ec64beca9f2c8848d15f9ad1835d7702390c13ce#diff-35b384f5686672e9b726b991a0c45dd8c18cfc8b365953489bf95a39db82205f
- https://github.com/rectorphp/rector-src/pull/5167/commits/ec64beca9f2c8848d15f9ad1835d7702390c13ce#diff-1d332eed33574b56a858a67eac5ef2d6d140534b69278e6c5d9353fe8b9cedb9

I think these example show the actual value of the rule

**Comment by staabm** [comment]
Anything I need todo to move this forward?

**Comment by TomasVotruba** [comment]
The rule purpose is unclear to me. The provided changes in Rector make code sligtly confusing, as value is dropped and null is used instead. The value is important for code understanding there.

Saying that, I probably won't merge this to Rector as rule should be clear and easy to understand.

**Comment by staabm** [comment]
The idea of the rule is: it simplifies the logic by replacing variables with constant values, when variables used in a return expression can only contain a constant value.

This usually is the case because of type substraction, e.g. when `true` was substracted from a `bool`, the remaining value can only be `false`. Another example is, when from `?A` the `null` get substracted, the remaining part only can be `A`.

-> the resulting code is more clear, because it no longer uses variables in return statements, where the variable itself can only contain a constant value

**Comment by TomasVotruba** [comment]
I understand it standardizes values to constant alternatives. Yet it takes away the readability and makes further variable changed or extension impossible.  

Saying that, this won't part of Rector set, as our goal is keeping code both high quality, and human-readable. 
This is better to add to a custom rules set. Thanks for understanding

**Pull request closed without merging**